fast TypeMeta/ScalarType conversion

bhosmer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #45263 use bitfield to shrink TensorImpl
* #45046 faster TensorOptions merging
* #45043 track Half/ComplexHalf default dtype
* #45026 pass TypeMeta by value
* **#44965 fast TypeMeta/ScalarType conversion**

Implements fast conversion between TypeMeta and ScalarType by having TypeMeta wrap only an index, which is equal to ScalarType enums for corresponding types. This simplifies a bunch of logic and logistics, e.g. making TypeMeta/ScalarType conversions basically bounds checks. Also it shrinks TypeMeta enough to get TensorImpl packed into 240 bits instead of 248.

Perf bump is good, both on FB microbenchmarks (FB folks see see Phabricator diff) and callgrind - here's callgrind dumps from [before](https://gist.github.com/bhosmer/1e0131b8516154a34f557cd29b6e7294) and [after](https://gist.github.com/bhosmer/894c84d65172dfe5b93db06d46102679) this stack. TLDR, ~9% fewer instructions on an `at::empty()` loop:

Before:
```
4,400,396,277 (100.0%)  PROGRAM TOTALS
```
After:
```
4,031,245,250 (100.0%)  PROGRAM TOTALS
```
code:
```
int main() {
  c10::IntArrayRef size{1};
  auto t = torch::empty(size);
  CALLGRIND_START_INSTRUMENTATION;
  for (int i = 0; i < 1000000; ++i) {
    t = at::empty(size);
  }
  CALLGRIND_STOP_INSTRUMENTATION;
  return 0;
}
```
Unfortunately I didn't keep callgrinds for the intermediate steps, but they're easy to repro. @ljk53's excellent setup is [here](https://github.com/ljk53/pytorch-tests/blob/master/benchmark/callgrind.md).

**[updated]** Future work: eliminate the redundancy between `TypeMetaData.id_` and `TypeMeta.index_`, either by storing index values as type ids in `TypeMetaData`, or removing the `id_` field entirely, if it turns out that all clients of this subsystem access TypeMetaData objects via TypeMeta, which can supply the index directly.

Review notes:
1. diff is a bit noisy, sorry - most is code movement due to header splitting for dependency management. The heart of the matter is `typeid.h` and `typeid.cpp`. 
2. **[updated]** adopted @ezyang's suggestion for fixing the original's read/write concurrency problem: a fixed-length array stores all `TypeMetaData` instances, imposing a limit on the total number of types that can be registered with `CAFFE_KNOWN_TYPE()`. That limit is set here at 256, but as before we use `uint16_t` to store index values, giving us wiggle room if that limit is too stingy. (There are currently no more than 70 calls to CAFFE_KNOWN_TYPE in master, and 17 ScalarType types, so we're about 1/3 full.)
3. A speedup tactic with the among the best ROI here was **pulling TORCH_CHECK and friends out of inline functions.** Looking at callgrind reaffirms the compiler's reluctance to inline a function with a throw in the body - just push the check into a helper. We should be on the lookout for the `inline foo() { TORCH_CHECK(...) }` pattern throughout the codebase.

Differential Revision: [D23789657](https://our.internmc.facebook.com/intern/diff/D23789657)

Commit:
[wip] fast typeMeta/ScalarType conversion approach 2

[ghstack-poisoned]

Commit:
Update on "[wip] fast typeMeta/ScalarType conversion approach 2"


Differential Revision: [D23789657](https://our.internmc.facebook.com/intern/diff/D23789657)

[ghstack-poisoned]

Commit:
Update on "[wip] fast typeMeta/ScalarType conversion approach 2"


Differential Revision: [D23789657](https://our.internmc.facebook.com/intern/diff/D23789657)

[ghstack-poisoned]

Commit:
Update on "[wip] fast typeMeta/ScalarType conversion approach 2"


Differential Revision: [D23789657](https://our.internmc.facebook.com/intern/diff/D23789657)

[ghstack-poisoned]

Commit:
Update on "[wip] fast typeMeta/ScalarType conversion approach 2"


Differential Revision: [D23789657](https://our.internmc.facebook.com/intern/diff/D23789657)

[ghstack-poisoned]

Commit:
Update on "[wip] fast typeMeta/ScalarType conversion approach 2"


Differential Revision: [D23789657](https://our.internmc.facebook.com/intern/diff/D23789657)

[ghstack-poisoned]

Commit:
Update on "[wip] fast typeMeta/ScalarType conversion approach 2"


Differential Revision: [D23789657](https://our.internmc.facebook.com/intern/diff/D23789657)

[ghstack-poisoned]

Commit:
Update on "fast TypeMeta/ScalarType conversion"


Implements fast conversion between TypeMeta and ScalarType by having TypeMeta wrap only an index, which is equal to ScalarType enums for corresponding types. This simplifies a bunch of logic and logistics, e.g. making TypeMeta/ScalarType conversions basically bounds checks. Also it shrinks TypeMeta enough to get TensorImpl packed into 240 bits instead of 248.

Perf bump is good, both on FB microbenchmarks (FB folks see see Phabricator diff) and callgrind - here's callgrind dumps from [before](https://gist.github.com/bhosmer/1e0131b8516154a34f557cd29b6e7294) and [after](https://gist.github.com/bhosmer/bce1816827122bf1646b9ff9dd69e1b6) this stack. TLDR:

Before:
```
4,400,396,277 (100.0%)  PROGRAM TOTALS
```
After:
```
4,029,396,354 (100.0%)  PROGRAM TOTALS
```
on this code:
```
int main() {
  c10::IntArrayRef size{1};
  auto t = torch::empty(size);
  CALLGRIND_START_INSTRUMENTATION;
  for (int i = 0; i < 1000000; ++i) {
    t = at::empty(size);
  }
  CALLGRIND_STOP_INSTRUMENTATION;
  return 0;
}
```
Unfortunately I didn't keep callgrinds for the intermediate steps, but they're easy to repro. @ljk53's excellent setup is [here](https://github.com/ljk53/pytorch-tests/blob/master/benchmark/callgrind.md).

Review notes:
1. diff is a bit noisy, sorry - most is code movement due to header splitting for dependency management. The heart of the matter is `typeid.h` and `typeid.cpp`. 
2. in there, in addition to eyeballing for SIOF issues, note the code that manages the vector of non-ScalarType TypeMetaData instances. This grows dynamically each time a TypeMeta<T> is first requested for a given T, so needs to be thread safe on updates, not just initialization. (Comments in code should clarify - if not, please note in review comments 😁 )
3. A speedup tactic with the among the best ROI here was **pulling TORCH_CHECK and friends out of inline functions.** Looking at callgrind reaffirms the compiler's reluctance to inline a function with a throw in the body - just push the check into a helper. We should be on the lookout for the `inline foo() { TORCH_CHECK(...) }` pattern throughout the codebase.

Differential Revision: [D23789657](https://our.internmc.facebook.com/intern/diff/D23789657)

[ghstack-poisoned]

ezyang commented on torch/csrc/jit/tensorexpr/stmt.h:
```diff
@@ -3,6 +3,7 @@
 #include <algorithm>
 #include <list>
 #include <string>
+#include <unordered_set>
```
What's going on here?

ezyang replied:
Oh this is because you are no longer recursively including typeid.h? I have always found that I regretted doing header changes in PRs, because there's always more to fix in fbcode later. If you run into more resistance you should probably consider undoing the header change.

bhosmer replied:
The change lets the TypeMeta stuff depend directly on ScalarType, but yeah, I could substitute some new transitive inclusion if it jitters fbcode too much.

ezyang commented on c10/util/typeid.h:
```diff
@@ -293,25 +291,25 @@ inline constexpr TypeMetaData::Delete* _PickDelete() noexcept {
+// with scalarTypeItemSizes as a constexpr static member used by
+// a public inline instance method
+//
+static constexpr uint16_t NumScalarTypes = static_cast<uint16_t>(ScalarType::NumOptions);
```
Since this is not in a class anymore, I'm pretty sure you don't need static anymore?

bhosmer replied:
Ah yeah, copy paste, thanks!

ezyang commented on c10/util/typeid.h:
```diff
@@ -442,29 +458,115 @@ class C10_API TypeMeta final {
+  }
+
+private:
+  static void error_unrecognized_scalartype(ScalarType scalar_type) {
```
Use `[[noreturn]]` to mark this as noreturn and then avoid "control reaches end of non-void function" above.

ezyang commented on c10/util/typeid.h:
```diff
@@ -442,29 +458,115 @@ class C10_API TypeMeta final {
+  }
+  static void error_unsupported_typemeta(caffe2::TypeMeta dtype) {
+    TORCH_CHECK(false, "Unsupported TypeMeta in ATen: ", dtype, " (please report this error)");
+  }
```
Wasn't your intention to hide these in a cpp file?

bhosmer replied:
It's not necessary to facilitate inlining, if that's you mean - but yeah that's where they should be.

ezyang commented on c10/util/typeid.h:
```diff
@@ -442,29 +458,115 @@ class C10_API TypeMeta final {
+    // so we don't need to worry about the same type being added twice.
+    // But different threads might be adding different Ts concurrently.
+    // So while we don't need to test for the presence of T before
+    // appending, we do need to serialize the appends themselves.
```
Nuh uh, you can't do this. If you have a reader access to vector in one thread, and a writer access to vector in another, you need synchronization on both ends. Writer synchronization isn't enough.

ezyang replied:
Which, for obvious reasons, is going to be a big performance problem.

ezyang replied:
cc @smessmer

ezyang replied:
The big hazard is `instances` reallocating the buffer. So one possible way we could fix this is add a compile-time "maximum number of typeids" and then just allocate the buffer large enough. Then concurrent reads could be implemented in a lock free manner.

bhosmer replied:
Ugh yeah realized this with a sinking feeling while brushing my teeth this morning. I think I temporarily hallucinated eager initialization that didn't interleave with accesses... so TypeMetaData is 72 bytes, and a quick grep says there are currently no more than 71 CAFFE_KNOWN_TYPEs. 256 would be a little over 18k. Does that sound reasonable (in both senses)?

ezyang commented on c10/util/typeid.h:
```diff
@@ -442,29 +458,115 @@ class C10_API TypeMeta final {
+
+  uint16_t index_;
+
+  inline const detail::TypeMetaData& data() const {
```
The if statement here seems totally unnecessary. Why not just initialize `nonScalarTypeMetaDatas` to start with the scalar type metadatas? Then you don't need to if statement here at all.

ljk53 replied:
I was typing the same comment then seeing yours :)

bhosmer replied:
You mean just pad it out? Sure why not. They wouldn't even need to be the actual scalar type metadatas, just the equivalent number of blanks - that would prevent people from accessing the dynamic vector instead of the static array.

bhosmer replied:
The original motivation was to save the extra hop to the heap-allocated (and volatile) vector backing store, and also take advantage of the array being constexpr, back when I thought it could be made constexpr unconditionally. The latter didn't pan out, but the former was paying off pretty nicely - until the fast-path `itemsize()` took the heat off this call path. The current savings on the microbenchmarks are tiny, and it's hard to see this path heating up on other use cases, so probably not worth the complexity. Assuming the solution to the concurrency problem doesn't slow down the dynamic vector to the point where the payoff here again becomes interesting, I'll get rid of the static one.

ezyang replied:
I mean, if the vector is going to have fixed max size, you can just preallocate a big array and then you get the best of both worlds (since we can save the hop).

ezyang commented on c10/util/typeid.h:
```diff
@@ -411,7 +427,7 @@ class C10_API TypeMeta final {
 
   template <class T>
-  static C10_TYPENAME_CONSTEXPR c10::string_view TypeName() noexcept {
+  static c10::string_view TypeName() noexcept {
```
Err, why did constexpr get removed here?

bhosmer replied:
Because it's a conditional macro, e.g. it's [not actually constexpr for GCC < 9](https://github.com/pytorch/pytorch/blob/master/c10/util/TypeIndex.h#L16-L55). Since no fundamental optimizations can rely on it - for instance, that's why `scalarTypeMetadatas` needs to be a static local rather than top level - I thought it was better to forego the extra obfuscation, build modality and so on and wait until we can provide total coverage.

ezyang commented on c10/util/typeid.h:
```diff
@@ -442,29 +458,115 @@ class C10_API TypeMeta final {
+  */
+  static inline caffe2::TypeMeta fromScalarType(ScalarType scalar_type) {
+    const size_t index = static_cast<uint16_t>(scalar_type);
+    if (C10_LIKELY(index < NumScalarTypes)) {
```
I think you should only do this test in debug mode. Under ordinary circumstances, the `ScalarType` should be enforced, by C++ type system, to be in bounds. I agree it could be corrupted, but I don't think it's worth testing here.

bhosmer replied:
Makes sense to me - I basically transcribed the original here on the premise that it was testing a condition that arose in practice, but I agree, it'd be hard to trigger it

ezyang commented:
The diff looks plausible but the concurrency problem is a showstopper

ljk53 commented on c10/util/typeid.h:
```diff
@@ -343,60 +341,78 @@ class C10_API TypeMeta final {
-    return data_->itemsize_;
+  inline size_t itemsize() const noexcept {
+    if (C10_LIKELY(isScalarType())) {
+      return scalarTypeItemSizes[index_];
```
Curious: what's the top line IR count impact of the fast path compared to using the general data().itemsize_ below? What should be the general guideline? e.g.: whether to create a fast path for, say, copy()?

bhosmer replied:
Ugh that's a good question and I don't have the callgrind data saved for this change - informally, anything that looked like it approached ~1% caught my eye - basically anything on the first page or so of callgrind listings that seemed like it was unreasonably high IR. 

itemsize() is called within empty_cpu(), which is why I saw it and none of the others in the benchmarks I was running, but it seemed like a decent balance to stop there, rather than do fast paths for the others on general principles. I figured if we see evidence that particular others are hot we can always add fast paths for them.

ezyang replied:
I'm guessing if you reserve space in `data` for the itemsize table rather than indirecting through vector you can eliminate the necessity for this fastpath.

bhosmer replied:
Not sure if it's what you mean, but yeah: always getting itemsize from the TypeMetaData array entries is close enough now to demotivate the side table - only about 0.1% difference. Putting up an update with that simplification, but LMK if you meant something smarter.

bhosmer replied:
Interestingly, I see what appear to be consistently better numbers on our in-house microbenchmarks with the fast path `itemsize()` is left in. Noise makes the source of the difference hard to bisect, but the difference across the stack is notable. I think I'll leave the fast path in at least until we can use @robieta's stuff to bisect with a bit less manual labor :)

bhosmer commented:
@ezyang @ljk53 @dzhulgakov thanks for the comments! I'll put up a "maximum number of typeids" solution to the concurrency issue as soon as I get some more coding time, hopefully within the next day or two.

Commit:
Update on "fast TypeMeta/ScalarType conversion"


Implements fast conversion between TypeMeta and ScalarType by having TypeMeta wrap only an index, which is equal to ScalarType enums for corresponding types. This simplifies a bunch of logic and logistics, e.g. making TypeMeta/ScalarType conversions basically bounds checks. Also it shrinks TypeMeta enough to get TensorImpl packed into 240 bits instead of 248.

Perf bump is good, both on FB microbenchmarks (FB folks see see Phabricator diff) and callgrind - here's callgrind dumps from [before](https://gist.github.com/bhosmer/1e0131b8516154a34f557cd29b6e7294) and [after](https://gist.github.com/bhosmer/bce1816827122bf1646b9ff9dd69e1b6) this stack. TLDR:

Before:
```
4,400,396,277 (100.0%)  PROGRAM TOTALS
```
After:
```
4,029,396,354 (100.0%)  PROGRAM TOTALS
```
on this code:
```
int main() {
  c10::IntArrayRef size{1};
  auto t = torch::empty(size);
  CALLGRIND_START_INSTRUMENTATION;
  for (int i = 0; i < 1000000; ++i) {
    t = at::empty(size);
  }
  CALLGRIND_STOP_INSTRUMENTATION;
  return 0;
}
```
Unfortunately I didn't keep callgrinds for the intermediate steps, but they're easy to repro. @ljk53's excellent setup is [here](https://github.com/ljk53/pytorch-tests/blob/master/benchmark/callgrind.md).

Review notes:
1. diff is a bit noisy, sorry - most is code movement due to header splitting for dependency management. The heart of the matter is `typeid.h` and `typeid.cpp`. 
2. in there, in addition to eyeballing for SIOF issues, note the code that manages the vector of non-ScalarType TypeMetaData instances. This grows dynamically each time a TypeMeta<T> is first requested for a given T, so needs to be thread safe on updates, not just initialization. (Comments in code should clarify - if not, please note in review comments 😁 )
3. A speedup tactic with the among the best ROI here was **pulling TORCH_CHECK and friends out of inline functions.** Looking at callgrind reaffirms the compiler's reluctance to inline a function with a throw in the body - just push the check into a helper. We should be on the lookout for the `inline foo() { TORCH_CHECK(...) }` pattern throughout the codebase.

Differential Revision: [D23789657](https://our.internmc.facebook.com/intern/diff/D23789657)

[ghstack-poisoned]

Commit:
Update on "fast TypeMeta/ScalarType conversion"


Implements fast conversion between TypeMeta and ScalarType by having TypeMeta wrap only an index, which is equal to ScalarType enums for corresponding types. This simplifies a bunch of logic and logistics, e.g. making TypeMeta/ScalarType conversions basically bounds checks. Also it shrinks TypeMeta enough to get TensorImpl packed into 240 bits instead of 248.

Perf bump is good, both on FB microbenchmarks (FB folks see see Phabricator diff) and callgrind - here's callgrind dumps from [before](https://gist.github.com/bhosmer/1e0131b8516154a34f557cd29b6e7294) and [after](https://gist.github.com/bhosmer/bce1816827122bf1646b9ff9dd69e1b6) this stack. TLDR:

Before:
```
4,400,396,277 (100.0%)  PROGRAM TOTALS
```
After:
```
4,029,396,354 (100.0%)  PROGRAM TOTALS
```
on this code:
```
int main() {
  c10::IntArrayRef size{1};
  auto t = torch::empty(size);
  CALLGRIND_START_INSTRUMENTATION;
  for (int i = 0; i < 1000000; ++i) {
    t = at::empty(size);
  }
  CALLGRIND_STOP_INSTRUMENTATION;
  return 0;
}
```
Unfortunately I didn't keep callgrinds for the intermediate steps, but they're easy to repro. @ljk53's excellent setup is [here](https://github.com/ljk53/pytorch-tests/blob/master/benchmark/callgrind.md).

Review notes:
1. diff is a bit noisy, sorry - most is code movement due to header splitting for dependency management. The heart of the matter is `typeid.h` and `typeid.cpp`. 
2. in there, in addition to eyeballing for SIOF issues, note the code that manages the vector of non-ScalarType TypeMetaData instances. This grows dynamically each time a TypeMeta<T> is first requested for a given T, so needs to be thread safe on updates, not just initialization. (Comments in code should clarify - if not, please note in review comments 😁 )
3. A speedup tactic with the among the best ROI here was **pulling TORCH_CHECK and friends out of inline functions.** Looking at callgrind reaffirms the compiler's reluctance to inline a function with a throw in the body - just push the check into a helper. We should be on the lookout for the `inline foo() { TORCH_CHECK(...) }` pattern throughout the codebase.

Differential Revision: [D23789657](https://our.internmc.facebook.com/intern/diff/D23789657)

[ghstack-poisoned]

ezyang reviewed (approved):
This looks correct and we should land it.

Commit:
Update on "fast TypeMeta/ScalarType conversion"


Implements fast conversion between TypeMeta and ScalarType by having TypeMeta wrap only an index, which is equal to ScalarType enums for corresponding types. This simplifies a bunch of logic and logistics, e.g. making TypeMeta/ScalarType conversions basically bounds checks. Also it shrinks TypeMeta enough to get TensorImpl packed into 240 bits instead of 248.

Perf bump is good, both on FB microbenchmarks (FB folks see see Phabricator diff) and callgrind - here's callgrind dumps from [before](https://gist.github.com/bhosmer/1e0131b8516154a34f557cd29b6e7294) and [after](https://gist.github.com/bhosmer/894c84d65172dfe5b93db06d46102679) this stack. TLDR, ~9% fewer instructions on an `at::empty()` loop:

Before:
```
4,400,396,277 (100.0%)  PROGRAM TOTALS
```
After:
```
4,031,245,250 (100.0%)  PROGRAM TOTALS
```
code:
```
int main() {
  c10::IntArrayRef size{1};
  auto t = torch::empty(size);
  CALLGRIND_START_INSTRUMENTATION;
  for (int i = 0; i < 1000000; ++i) {
    t = at::empty(size);
  }
  CALLGRIND_STOP_INSTRUMENTATION;
  return 0;
}
```
Unfortunately I didn't keep callgrinds for the intermediate steps, but they're easy to repro. @ljk53's excellent setup is [here](https://github.com/ljk53/pytorch-tests/blob/master/benchmark/callgrind.md).

**[updated]** Future work: eliminate the redundancy between `TypeMetaData.id_` and `TypeMeta.index_`, either by storing index values as type ids in `TypeMetaData`, or removing the `id_` field entirely, if it turns out that all clients of this subsystem access TypeMetaData objects via TypeMeta, which can supply the index directly.

Review notes:
1. diff is a bit noisy, sorry - most is code movement due to header splitting for dependency management. The heart of the matter is `typeid.h` and `typeid.cpp`. 
2. **[updated]** adopted @ezyang's suggestion for fixing the original's read/write concurrency problem: a fixed-length array stores all `TypeMetaData` instances, imposing a limit on the total number of types that can be registered with `CAFFE_KNOWN_TYPE()`. That limit is set here at 256, but as before we use `uint16_t` to store index values, giving us wiggle room if that limit is too stingy. (There are currently no more than 70 calls to CAFFE_KNOWN_TYPE in master, and 17 ScalarType types, so we're about 1/3 full.)
3. A speedup tactic with the among the best ROI here was **pulling TORCH_CHECK and friends out of inline functions.** Looking at callgrind reaffirms the compiler's reluctance to inline a function with a throw in the body - just push the check into a helper. We should be on the lookout for the `inline foo() { TORCH_CHECK(...) }` pattern throughout the codebase.

Differential Revision: [D23789657](https://our.internmc.facebook.com/intern/diff/D23789657)

[ghstack-poisoned]

Commit:
Update on "fast TypeMeta/ScalarType conversion"


Implements fast conversion between TypeMeta and ScalarType by having TypeMeta wrap only an index, which is equal to ScalarType enums for corresponding types. This simplifies a bunch of logic and logistics, e.g. making TypeMeta/ScalarType conversions basically bounds checks. Also it shrinks TypeMeta enough to get TensorImpl packed into 240 bits instead of 248.

Perf bump is good, both on FB microbenchmarks (FB folks see see Phabricator diff) and callgrind - here's callgrind dumps from [before](https://gist.github.com/bhosmer/1e0131b8516154a34f557cd29b6e7294) and [after](https://gist.github.com/bhosmer/894c84d65172dfe5b93db06d46102679) this stack. TLDR, ~9% fewer instructions on an `at::empty()` loop:

Before:
```
4,400,396,277 (100.0%)  PROGRAM TOTALS
```
After:
```
4,031,245,250 (100.0%)  PROGRAM TOTALS
```
code:
```
int main() {
  c10::IntArrayRef size{1};
  auto t = torch::empty(size);
  CALLGRIND_START_INSTRUMENTATION;
  for (int i = 0; i < 1000000; ++i) {
    t = at::empty(size);
  }
  CALLGRIND_STOP_INSTRUMENTATION;
  return 0;
}
```
Unfortunately I didn't keep callgrinds for the intermediate steps, but they're easy to repro. @ljk53's excellent setup is [here](https://github.com/ljk53/pytorch-tests/blob/master/benchmark/callgrind.md).

**[updated]** Future work: eliminate the redundancy between `TypeMetaData.id_` and `TypeMeta.index_`, either by storing index values as type ids in `TypeMetaData`, or removing the `id_` field entirely, if it turns out that all clients of this subsystem access TypeMetaData objects via TypeMeta, which can supply the index directly.

Review notes:
1. diff is a bit noisy, sorry - most is code movement due to header splitting for dependency management. The heart of the matter is `typeid.h` and `typeid.cpp`. 
2. **[updated]** adopted @ezyang's suggestion for fixing the original's read/write concurrency problem: a fixed-length array stores all `TypeMetaData` instances, imposing a limit on the total number of types that can be registered with `CAFFE_KNOWN_TYPE()`. That limit is set here at 256, but as before we use `uint16_t` to store index values, giving us wiggle room if that limit is too stingy. (There are currently no more than 70 calls to CAFFE_KNOWN_TYPE in master, and 17 ScalarType types, so we're about 1/3 full.)
3. A speedup tactic with the among the best ROI here was **pulling TORCH_CHECK and friends out of inline functions.** Looking at callgrind reaffirms the compiler's reluctance to inline a function with a throw in the body - just push the check into a helper. We should be on the lookout for the `inline foo() { TORCH_CHECK(...) }` pattern throughout the codebase.

Differential Revision: [D23789657](https://our.internmc.facebook.com/intern/diff/D23789657)

[ghstack-poisoned]

Commit:
Update on "fast TypeMeta/ScalarType conversion"


Implements fast conversion between TypeMeta and ScalarType by having TypeMeta wrap only an index, which is equal to ScalarType enums for corresponding types. This simplifies a bunch of logic and logistics, e.g. making TypeMeta/ScalarType conversions basically bounds checks. Also it shrinks TypeMeta enough to get TensorImpl packed into 240 bits instead of 248.

Perf bump is good, both on FB microbenchmarks (FB folks see see Phabricator diff) and callgrind - here's callgrind dumps from [before](https://gist.github.com/bhosmer/1e0131b8516154a34f557cd29b6e7294) and [after](https://gist.github.com/bhosmer/894c84d65172dfe5b93db06d46102679) this stack. TLDR, ~9% fewer instructions on an `at::empty()` loop:

Before:
```
4,400,396,277 (100.0%)  PROGRAM TOTALS
```
After:
```
4,031,245,250 (100.0%)  PROGRAM TOTALS
```
code:
```
int main() {
  c10::IntArrayRef size{1};
  auto t = torch::empty(size);
  CALLGRIND_START_INSTRUMENTATION;
  for (int i = 0; i < 1000000; ++i) {
    t = at::empty(size);
  }
  CALLGRIND_STOP_INSTRUMENTATION;
  return 0;
}
```
Unfortunately I didn't keep callgrinds for the intermediate steps, but they're easy to repro. @ljk53's excellent setup is [here](https://github.com/ljk53/pytorch-tests/blob/master/benchmark/callgrind.md).

**[updated]** Future work: eliminate the redundancy between `TypeMetaData.id_` and `TypeMeta.index_`, either by storing index values as type ids in `TypeMetaData`, or removing the `id_` field entirely, if it turns out that all clients of this subsystem access TypeMetaData objects via TypeMeta, which can supply the index directly.

Review notes:
1. diff is a bit noisy, sorry - most is code movement due to header splitting for dependency management. The heart of the matter is `typeid.h` and `typeid.cpp`. 
2. **[updated]** adopted @ezyang's suggestion for fixing the original's read/write concurrency problem: a fixed-length array stores all `TypeMetaData` instances, imposing a limit on the total number of types that can be registered with `CAFFE_KNOWN_TYPE()`. That limit is set here at 256, but as before we use `uint16_t` to store index values, giving us wiggle room if that limit is too stingy. (There are currently no more than 70 calls to CAFFE_KNOWN_TYPE in master, and 17 ScalarType types, so we're about 1/3 full.)
3. A speedup tactic with the among the best ROI here was **pulling TORCH_CHECK and friends out of inline functions.** Looking at callgrind reaffirms the compiler's reluctance to inline a function with a throw in the body - just push the check into a helper. We should be on the lookout for the `inline foo() { TORCH_CHECK(...) }` pattern throughout the codebase.

Differential Revision: [D23789657](https://our.internmc.facebook.com/intern/diff/D23789657)

[ghstack-poisoned]

Commit:
Update on "fast TypeMeta/ScalarType conversion"


Implements fast conversion between TypeMeta and ScalarType by having TypeMeta wrap only an index, which is equal to ScalarType enums for corresponding types. This simplifies a bunch of logic and logistics, e.g. making TypeMeta/ScalarType conversions basically bounds checks. Also it shrinks TypeMeta enough to get TensorImpl packed into 240 bits instead of 248.

Perf bump is good, both on FB microbenchmarks (FB folks see see Phabricator diff) and callgrind - here's callgrind dumps from [before](https://gist.github.com/bhosmer/1e0131b8516154a34f557cd29b6e7294) and [after](https://gist.github.com/bhosmer/894c84d65172dfe5b93db06d46102679) this stack. TLDR, ~9% fewer instructions on an `at::empty()` loop:

Before:
```
4,400,396,277 (100.0%)  PROGRAM TOTALS
```
After:
```
4,031,245,250 (100.0%)  PROGRAM TOTALS
```
code:
```
int main() {
  c10::IntArrayRef size{1};
  auto t = torch::empty(size);
  CALLGRIND_START_INSTRUMENTATION;
  for (int i = 0; i < 1000000; ++i) {
    t = at::empty(size);
  }
  CALLGRIND_STOP_INSTRUMENTATION;
  return 0;
}
```
Unfortunately I didn't keep callgrinds for the intermediate steps, but they're easy to repro. @ljk53's excellent setup is [here](https://github.com/ljk53/pytorch-tests/blob/master/benchmark/callgrind.md).

**[updated]** Future work: eliminate the redundancy between `TypeMetaData.id_` and `TypeMeta.index_`, either by storing index values as type ids in `TypeMetaData`, or removing the `id_` field entirely, if it turns out that all clients of this subsystem access TypeMetaData objects via TypeMeta, which can supply the index directly.

Review notes:
1. diff is a bit noisy, sorry - most is code movement due to header splitting for dependency management. The heart of the matter is `typeid.h` and `typeid.cpp`. 
2. **[updated]** adopted @ezyang's suggestion for fixing the original's read/write concurrency problem: a fixed-length array stores all `TypeMetaData` instances, imposing a limit on the total number of types that can be registered with `CAFFE_KNOWN_TYPE()`. That limit is set here at 256, but as before we use `uint16_t` to store index values, giving us wiggle room if that limit is too stingy. (There are currently no more than 70 calls to CAFFE_KNOWN_TYPE in master, and 17 ScalarType types, so we're about 1/3 full.)
3. A speedup tactic with the among the best ROI here was **pulling TORCH_CHECK and friends out of inline functions.** Looking at callgrind reaffirms the compiler's reluctance to inline a function with a throw in the body - just push the check into a helper. We should be on the lookout for the `inline foo() { TORCH_CHECK(...) }` pattern throughout the codebase.

Differential Revision: [D23789657](https://our.internmc.facebook.com/intern/diff/D23789657)

[ghstack-poisoned]

Commit:
Update on "fast TypeMeta/ScalarType conversion"


Implements fast conversion between TypeMeta and ScalarType by having TypeMeta wrap only an index, which is equal to ScalarType enums for corresponding types. This simplifies a bunch of logic and logistics, e.g. making TypeMeta/ScalarType conversions basically bounds checks. Also it shrinks TypeMeta enough to get TensorImpl packed into 240 bits instead of 248.

Perf bump is good, both on FB microbenchmarks (FB folks see see Phabricator diff) and callgrind - here's callgrind dumps from [before](https://gist.github.com/bhosmer/1e0131b8516154a34f557cd29b6e7294) and [after](https://gist.github.com/bhosmer/894c84d65172dfe5b93db06d46102679) this stack. TLDR, ~9% fewer instructions on an `at::empty()` loop:

Before:
```
4,400,396,277 (100.0%)  PROGRAM TOTALS
```
After:
```
4,031,245,250 (100.0%)  PROGRAM TOTALS
```
code:
```
int main() {
  c10::IntArrayRef size{1};
  auto t = torch::empty(size);
  CALLGRIND_START_INSTRUMENTATION;
  for (int i = 0; i < 1000000; ++i) {
    t = at::empty(size);
  }
  CALLGRIND_STOP_INSTRUMENTATION;
  return 0;
}
```
Unfortunately I didn't keep callgrinds for the intermediate steps, but they're easy to repro. @ljk53's excellent setup is [here](https://github.com/ljk53/pytorch-tests/blob/master/benchmark/callgrind.md).

**[updated]** Future work: eliminate the redundancy between `TypeMetaData.id_` and `TypeMeta.index_`, either by storing index values as type ids in `TypeMetaData`, or removing the `id_` field entirely, if it turns out that all clients of this subsystem access TypeMetaData objects via TypeMeta, which can supply the index directly.

Review notes:
1. diff is a bit noisy, sorry - most is code movement due to header splitting for dependency management. The heart of the matter is `typeid.h` and `typeid.cpp`. 
2. **[updated]** adopted @ezyang's suggestion for fixing the original's read/write concurrency problem: a fixed-length array stores all `TypeMetaData` instances, imposing a limit on the total number of types that can be registered with `CAFFE_KNOWN_TYPE()`. That limit is set here at 256, but as before we use `uint16_t` to store index values, giving us wiggle room if that limit is too stingy. (There are currently no more than 70 calls to CAFFE_KNOWN_TYPE in master, and 17 ScalarType types, so we're about 1/3 full.)
3. A speedup tactic with the among the best ROI here was **pulling TORCH_CHECK and friends out of inline functions.** Looking at callgrind reaffirms the compiler's reluctance to inline a function with a throw in the body - just push the check into a helper. We should be on the lookout for the `inline foo() { TORCH_CHECK(...) }` pattern throughout the codebase.

Differential Revision: [D23789657](https://our.internmc.facebook.com/intern/diff/D23789657)

[ghstack-poisoned]

Commit:
Update on "fast TypeMeta/ScalarType conversion"


Implements fast conversion between TypeMeta and ScalarType by having TypeMeta wrap only an index, which is equal to ScalarType enums for corresponding types. This simplifies a bunch of logic and logistics, e.g. making TypeMeta/ScalarType conversions basically bounds checks. Also it shrinks TypeMeta enough to get TensorImpl packed into 240 bits instead of 248.

Perf bump is good, both on FB microbenchmarks (FB folks see see Phabricator diff) and callgrind - here's callgrind dumps from [before](https://gist.github.com/bhosmer/1e0131b8516154a34f557cd29b6e7294) and [after](https://gist.github.com/bhosmer/894c84d65172dfe5b93db06d46102679) this stack. TLDR, ~9% fewer instructions on an `at::empty()` loop:

Before:
```
4,400,396,277 (100.0%)  PROGRAM TOTALS
```
After:
```
4,031,245,250 (100.0%)  PROGRAM TOTALS
```
code:
```
int main() {
  c10::IntArrayRef size{1};
  auto t = torch::empty(size);
  CALLGRIND_START_INSTRUMENTATION;
  for (int i = 0; i < 1000000; ++i) {
    t = at::empty(size);
  }
  CALLGRIND_STOP_INSTRUMENTATION;
  return 0;
}
```
Unfortunately I didn't keep callgrinds for the intermediate steps, but they're easy to repro. @ljk53's excellent setup is [here](https://github.com/ljk53/pytorch-tests/blob/master/benchmark/callgrind.md).

**[updated]** Future work: eliminate the redundancy between `TypeMetaData.id_` and `TypeMeta.index_`, either by storing index values as type ids in `TypeMetaData`, or removing the `id_` field entirely, if it turns out that all clients of this subsystem access TypeMetaData objects via TypeMeta, which can supply the index directly.

Review notes:
1. diff is a bit noisy, sorry - most is code movement due to header splitting for dependency management. The heart of the matter is `typeid.h` and `typeid.cpp`. 
2. **[updated]** adopted @ezyang's suggestion for fixing the original's read/write concurrency problem: a fixed-length array stores all `TypeMetaData` instances, imposing a limit on the total number of types that can be registered with `CAFFE_KNOWN_TYPE()`. That limit is set here at 256, but as before we use `uint16_t` to store index values, giving us wiggle room if that limit is too stingy. (There are currently no more than 70 calls to CAFFE_KNOWN_TYPE in master, and 17 ScalarType types, so we're about 1/3 full.)
3. A speedup tactic with the among the best ROI here was **pulling TORCH_CHECK and friends out of inline functions.** Looking at callgrind reaffirms the compiler's reluctance to inline a function with a throw in the body - just push the check into a helper. We should be on the lookout for the `inline foo() { TORCH_CHECK(...) }` pattern throughout the codebase.

Differential Revision: [D23789657](https://our.internmc.facebook.com/intern/diff/D23789657)

[ghstack-poisoned]

Commit:
Update on "fast TypeMeta/ScalarType conversion"


Implements fast conversion between TypeMeta and ScalarType by having TypeMeta wrap only an index, which is equal to ScalarType enums for corresponding types. This simplifies a bunch of logic and logistics, e.g. making TypeMeta/ScalarType conversions basically bounds checks. Also it shrinks TypeMeta enough to get TensorImpl packed into 240 bits instead of 248.

Perf bump is good, both on FB microbenchmarks (FB folks see see Phabricator diff) and callgrind - here's callgrind dumps from [before](https://gist.github.com/bhosmer/1e0131b8516154a34f557cd29b6e7294) and [after](https://gist.github.com/bhosmer/894c84d65172dfe5b93db06d46102679) this stack. TLDR, ~9% fewer instructions on an `at::empty()` loop:

Before:
```
4,400,396,277 (100.0%)  PROGRAM TOTALS
```
After:
```
4,031,245,250 (100.0%)  PROGRAM TOTALS
```
code:
```
int main() {
  c10::IntArrayRef size{1};
  auto t = torch::empty(size);
  CALLGRIND_START_INSTRUMENTATION;
  for (int i = 0; i < 1000000; ++i) {
    t = at::empty(size);
  }
  CALLGRIND_STOP_INSTRUMENTATION;
  return 0;
}
```
Unfortunately I didn't keep callgrinds for the intermediate steps, but they're easy to repro. @ljk53's excellent setup is [here](https://github.com/ljk53/pytorch-tests/blob/master/benchmark/callgrind.md).

**[updated]** Future work: eliminate the redundancy between `TypeMetaData.id_` and `TypeMeta.index_`, either by storing index values as type ids in `TypeMetaData`, or removing the `id_` field entirely, if it turns out that all clients of this subsystem access TypeMetaData objects via TypeMeta, which can supply the index directly.

Review notes:
1. diff is a bit noisy, sorry - most is code movement due to header splitting for dependency management. The heart of the matter is `typeid.h` and `typeid.cpp`. 
2. **[updated]** adopted @ezyang's suggestion for fixing the original's read/write concurrency problem: a fixed-length array stores all `TypeMetaData` instances, imposing a limit on the total number of types that can be registered with `CAFFE_KNOWN_TYPE()`. That limit is set here at 256, but as before we use `uint16_t` to store index values, giving us wiggle room if that limit is too stingy. (There are currently no more than 70 calls to CAFFE_KNOWN_TYPE in master, and 17 ScalarType types, so we're about 1/3 full.)
3. A speedup tactic with the among the best ROI here was **pulling TORCH_CHECK and friends out of inline functions.** Looking at callgrind reaffirms the compiler's reluctance to inline a function with a throw in the body - just push the check into a helper. We should be on the lookout for the `inline foo() { TORCH_CHECK(...) }` pattern throughout the codebase.

Differential Revision: [D23789657](https://our.internmc.facebook.com/intern/diff/D23789657)

[ghstack-poisoned]

Commit:
Update on "fast TypeMeta/ScalarType conversion"


Implements fast conversion between TypeMeta and ScalarType by having TypeMeta wrap only an index, which is equal to ScalarType enums for corresponding types. This simplifies a bunch of logic and logistics, e.g. making TypeMeta/ScalarType conversions basically bounds checks. Also it shrinks TypeMeta enough to get TensorImpl packed into 240 bits instead of 248.

Perf bump is good, both on FB microbenchmarks (FB folks see see Phabricator diff) and callgrind - here's callgrind dumps from [before](https://gist.github.com/bhosmer/1e0131b8516154a34f557cd29b6e7294) and [after](https://gist.github.com/bhosmer/894c84d65172dfe5b93db06d46102679) this stack. TLDR, ~9% fewer instructions on an `at::empty()` loop:

Before:
```
4,400,396,277 (100.0%)  PROGRAM TOTALS
```
After:
```
4,031,245,250 (100.0%)  PROGRAM TOTALS
```
code:
```
int main() {
  c10::IntArrayRef size{1};
  auto t = torch::empty(size);
  CALLGRIND_START_INSTRUMENTATION;
  for (int i = 0; i < 1000000; ++i) {
    t = at::empty(size);
  }
  CALLGRIND_STOP_INSTRUMENTATION;
  return 0;
}
```
Unfortunately I didn't keep callgrinds for the intermediate steps, but they're easy to repro. @ljk53's excellent setup is [here](https://github.com/ljk53/pytorch-tests/blob/master/benchmark/callgrind.md).

**[updated]** Future work: eliminate the redundancy between `TypeMetaData.id_` and `TypeMeta.index_`, either by storing index values as type ids in `TypeMetaData`, or removing the `id_` field entirely, if it turns out that all clients of this subsystem access TypeMetaData objects via TypeMeta, which can supply the index directly.

Review notes:
1. diff is a bit noisy, sorry - most is code movement due to header splitting for dependency management. The heart of the matter is `typeid.h` and `typeid.cpp`. 
2. **[updated]** adopted @ezyang's suggestion for fixing the original's read/write concurrency problem: a fixed-length array stores all `TypeMetaData` instances, imposing a limit on the total number of types that can be registered with `CAFFE_KNOWN_TYPE()`. That limit is set here at 256, but as before we use `uint16_t` to store index values, giving us wiggle room if that limit is too stingy. (There are currently no more than 70 calls to CAFFE_KNOWN_TYPE in master, and 17 ScalarType types, so we're about 1/3 full.)
3. A speedup tactic with the among the best ROI here was **pulling TORCH_CHECK and friends out of inline functions.** Looking at callgrind reaffirms the compiler's reluctance to inline a function with a throw in the body - just push the check into a helper. We should be on the lookout for the `inline foo() { TORCH_CHECK(...) }` pattern throughout the codebase.

Differential Revision: [D23789657](https://our.internmc.facebook.com/intern/diff/D23789657)

[ghstack-poisoned]

mruberry commented:
Unlanding. This appears to have broken a ROCm build. Relevant log snippet:

```
00:31:41 test_autocast_torch_fp16 (__main__.TestCuda) ... MIOpen Error: /root/driver/MLOpen/src/gemm_v2.cpp:523: rocBlas error encountered
00:31:42 ERROR
00:31:42 test_autocast_torch_fp32 (__main__.TestCuda) ... Memory exception on virtual address 0x0, node id 4 : Page not present
00:31:42 Address does not belong to a known buffer
00:31:42 Memory access fault by GPU node-4 (Agent handle: 0x55f8887583c0) on address (nil). Reason: Page not present or supervisor privilege.
00:31:42 Traceback (most recent call last):
00:31:42   File "test/run_test.py", line 746, in <module>
00:31:42     main()
00:31:42   File "test/run_test.py", line 729, in main
00:31:42     raise RuntimeError(err_message)
00:31:42 RuntimeError: test_cuda failed! Received signal: SIGIOT
```
